onnxruntime/core/common/threadpool.cc

```diff
@@ -21,12 +21,222 @@ limitations under the License.
+  }
+}
+
+void ThreadPoolProfiler::LogEnd(ThreadPoolEvent evt) {
+  if (enabled_) {
+    GetMainThreadStat().LogEnd(evt);
+  }
+}
+
+void ThreadPoolProfiler::LogEndAndStart(ThreadPoolEvent evt) {
+  if (enabled_) {
+    GetMainThreadStat().LogEndAndStart(evt);
+  }
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogCore() {
+#ifdef _WIN32
+  core_ = GetCurrentProcessorNumber();
+#elif defined(__APPLE__)
+  uint32_t CPUInfo[4];
+  __cpuid_count(1, 0, CPUInfo[0], CPUInfo[1], CPUInfo[2], CPUInfo[3]);
+  if ((CPUInfo[3] & (1 << 9)) != 0) {
+    core_ = (unsigned)CPUInfo[1] >> 24;
+  }
+#else
+  core_ = sched_getcpu();
+#endif
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogBlockSize(std::ptrdiff_t block_size) {
+  blocks_.emplace_back(block_size);
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogStart() {
+  points_.emplace_back(Clock::now());
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogEnd(ThreadPoolEvent evt) {
+  ORT_ENFORCE(!points_.empty(), "LogStart must pair with LogEnd");
+  events_[evt] += TimeDiffMicroSeconds(points_.back(), Clock::now());
+  points_.pop_back();
+}
+
+void ThreadPoolProfiler::MainThreadStat::LogEndAndStart(ThreadPoolEvent evt) {
+  ORT_ENFORCE(!points_.empty(), "LogStart must pair with LogEnd");
+  events_[evt] += TimeDiffMicroSeconds(points_.back(), Clock::now());
+  points_.back() = Clock::now();
+}
+
+std::string ThreadPoolProfiler::MainThreadStat::Reset() {
+  ORT_ENFORCE(points_.empty(), "LogStart must pair with LogEnd");
+  std::stringstream ss;
+  ss << "\"thread_id\": \"" << std::this_thread::get_id() << "\", \"block_size\": [";
+  if (!blocks_.empty()) {
+    std::copy(blocks_.begin(), blocks_.end() - 1, std::ostream_iterator<std::ptrdiff_t>(ss, ", "));
+    ss << blocks_.back();
+    blocks_.clear();
+  }
+  ss << "], \"core\": " << core_ << ", ";
+  for (int i = 0; i < MAX_EVENT; ++i) {
+    ss << "\"" << ThreadPoolProfiler::GetEventName(static_cast<ThreadPoolEvent>(i))
+       << "\": " << events_[i] << ((i == MAX_EVENT - 1) ? std::string{} : ", ");
+  }
+  memset(events_, 0, sizeof(uint64_t) * MAX_EVENT);
+  return ss.str();
+}
+
+const char* ThreadPoolProfiler::GetEventName(ThreadPoolEvent event) {
+  switch (event) {
+    case DISTRIBUTION:
+      return "Distribution";
+    case DISTRIBUTION_ENQUEUE:
+      return "DistributionEnqueue";
+    case RUN:
+      return "Run";
+    case WAIT:
+      return "Wait";
+    case WAIT_REVOKE:
+      return "WaitRevoke";
+    default:
+      return "UnknownEvent";
+  }
+}
+
+void ThreadPoolProfiler::LogThreadId(int thread_idx) {
+  child_thread_stats_[thread_idx].thread_id_ = std::this_thread::get_id();
+}
+
+void ThreadPoolProfiler::LogRun(int thread_idx) {
+  if (enabled_) {
+    child_thread_stats_[thread_idx].num_run_++;
+  }
+}
+
+void ThreadPoolProfiler::LogCore(int thread_idx) {
+  if (enabled_) {
+    auto now = Clock::now();
+    if (TimeDiffMicroSeconds(now, child_thread_stats_[thread_idx].last_logged_point_) > 1000) {
```

Comment:
This will be 1ms, so I am concerned about the overhead of collecting data here -- both in terms of getting the current time, and then in getting the current CPU (which may require a system call on some OS, although it has often been optimized).  We need to either check experimentally that there is no overhead, or be more conservative here (say, 100ms).

Reply:
From Polaris test we know that each request cost between 2 to 5 ms,
so how about let it be 10 ms per sampling?

Reply:
Reset to 10ms